vortex-layout/src/layouts/stats/builder.rs

```diff
@@ -0,0 +1,306 @@
+            Box::new(BoolBuilder::with_capacity(*n, capacity)),
+        )),
+        DType::Primitive(ptype, n) => Box::new(NamedArrayBuilder::new(
+            stat,
+            match_each_native_ptype!(ptype, |$P| {
+                Box::new(PrimitiveBuilder::<$P>::with_capacity(*n, capacity))
+            }),
+        )),
+        DType::Decimal(decimal_type, n) => Box::new(NamedArrayBuilder::new(
+            stat,
+            match_each_decimal_value_type!(precision_to_storage_size(decimal_type), |$D| {
+                Box::new(DecimalBuilder::<$D>::with_capacity(capacity, decimal_type.clone(), *n))
+            }),
+        )),
+        d @ DType::Utf8(_) | d @ DType::Binary(_) => match stat {
+            Stat::Max => Box::new(TruncatedMaxBinaryStatsBuilder::<Utf8Scalar>::new(
+                VarBinViewBuilder::with_capacity(d.clone(), capacity),
+                BoolBuilder::with_capacity(Nullability::NonNullable, capacity),
+                max_length,
+            )),
+            Stat::Min => Box::new(TruncatedMinBinaryStatsBuilder::<Utf8Scalar>::new(
+                VarBinViewBuilder::with_capacity(d.clone(), capacity),
+                BoolBuilder::with_capacity(Nullability::NonNullable, capacity),
+                max_length,
+            )),
+            _ => Box::new(NamedArrayBuilder::new(
+                stat,
+                Box::new(VarBinViewBuilder::with_capacity(d.clone(), capacity)),
+            )),
+        },
+        DType::Struct(struct_dtype, n) => Box::new(NamedArrayBuilder::new(
+            stat,
+            Box::new(StructBuilder::with_capacity(
+                struct_dtype.clone(),
+                *n,
+                capacity,
+            )),
+        )),
+        DType::List(dtype, n) => Box::new(NamedArrayBuilder::new(
+            stat,
+            Box::new(ListBuilder::<u64>::with_capacity(
+                dtype.clone(),
+                *n,
+                capacity,
+            )),
+        )),
+        DType::Extension(ext_dtype) => Box::new(NamedArrayBuilder::new(
+            stat,
+            Box::new(ExtensionBuilder::with_capacity(ext_dtype.clone(), capacity)),
+        )),
+    }
+}
+
+/// Arrays with their associated names, reduced version of a StructArray
+pub struct NamedArrays {
+    pub names: Vec<FieldName>,
+    pub arrays: Vec<ArrayRef>,
+}
+
+impl NamedArrays {
+    pub fn all_invalid(&self) -> VortexResult<bool> {
+        // by convention we assume that the first array is the one we care about for logical validity
+        self.arrays[0].all_invalid()
+    }
+}
+
+/// Minimal array builder interface for use by StatsTable for building stats arrays
+pub trait StatsArrayBuilder: Send {
+    fn stat(&self) -> Stat;
+
+    fn append_scalar_value(&mut self, value: ScalarValue) -> VortexResult<()>;
+
+    fn append_null(&mut self);
+
+    fn finish(&mut self) -> NamedArrays;
+}
+
+pub struct NamedArrayBuilder {
+    stat: Stat,
+    builder: Box<dyn ArrayBuilder>,
+}
```

Comment:
Could avoid double boxing if we were to specialize this for all primitive builder types